0001-scripting.md

```diff
@@ -0,0 +1,144 @@
+
+* Creating a new component type at runtime could be handled by having a special non-generic storage type that would store `Vec<u8>`. Those `Vec<u8>` would contain the raw data of the new components. Then, those components would be exposed to the scripting runtime through newly generated C headers corresponding to the component schema the user provided. That however will mean that TypeId can no longer be the only thing determining ResourceId, as those storages would share the same type but a different ID. But it does not seem unreasonable to have ResourceId become a `(TypeId, InstanceId)`, at least from my uneducated point of view. Also, at the end of the execution, the dynamic component should be collapsed into a statically generated component to no longer need to store the data in a `Vec<u8>`. The typical wokflow would be: create your components on the fly, do your testing with it, and if it's too slow then reboot the game. It seems reasonable to me, and besides I don't really expect it to be that slow, especially if you're only using it for logic creation.
+* Modifying existing component types on the fly would only work on component types that were created from the scripting environment, but I think it's a reasonable constraint. Modifying a component in dynamic state would simply be a matter of iterating over the `Vec<u8>` storage and moving the data inside around. Modifying a component in statically-collapsed state however is a bit trickier, but not that hard either. You need to replace all static instances of that component into a dynamic instance of the new component.
+* Removing an existing component type at least is easy. Just remove all instances of it and leave it all alone, scared, in a white void up to the next reboot where you end its sufferings.
+
+On the matter of resources, they simply can behave the same way component storages do.
+
+As specified, resources and components would declared using a schema language (the exact language is yet to be determined), offering the user numerous primitives such as different number types, strings, vectors, algebra primitives, etc. That way, the user can represent complex data structures from the schema. All those primitives must be sufficiently represented with C types so that they can be shared by all drivers. One important aspect is that, when possible, they do not hold references.
+
+A user can potentially force a reference to be stored in a component of a resource. But this is out of the scope of the scripting support: local safety has already been traded off by not using Rust. We only need to ensure the scripting languages can not disrupt the safety of code outside of the scripting world.
+
+## Satisfaction of goals
+
+Let us now take every goals one by one and explain how this proposal solves them specifically.
+
+* Support any language: a language driver can be written for any programming language that supports C FFI, which corresponds to any serious programming language available.
```

Comment:
I think this is a very broad statement. I would suggest re-wording it as a list of requirements for a language to be compatible, a C FFI being one of them.

Will there be some number of languages whose drivers are officially endorsed and maintained by the scripting team?

Reply:
I mention in the RFC that we should at least maintain LuaJIT and Rust.
We can also bring other drivers in the organization if other people want to build them. As described in the RFC, maintaining a language driver requires minimal work. The tricky part is designing the driver itself.